plugins/zify/zify.ml

```diff
@@ -0,0 +1,1095 @@
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+open Constr
+open Names
+open Pp
+open Lazy
+
+(** [get_type_of] performs beta reduction ;
+    Is it ok for Retyping.get_type_of (Zpower_nat n q) to return (fun _ : nat => Z) q ? *)
+let get_type_of env evd e =
+  Tacred.cbv_beta env evd (Retyping.get_type_of env evd e)
+
+(** [unsafe_to_constr c] returns a [Constr.t] without considering an evar_map.
+    This is useful for calling Constr.hash *)
+let unsafe_to_constr = EConstr.Unsafe.to_constr
+
+let pr_constr env evd e =
+  Printer.pr_constr_env env evd (unsafe_to_constr e)
+
+
+let debug = false (** Debug flag to get some intermediate terms *)
+
+(** [get_arrow_typ evd t] returns [t1;.tn] such that t = t1 -> .. -> tn.ci_npar
+    (only syntactic matching)
+ *)
+let rec get_arrow_typ evd t  =
+  match EConstr.kind evd t with
+  | Prod (a,p1,p2) (*when a.Context.binder_name = Names.Anonymous*) ->
+       p1::(get_arrow_typ evd p2)
+  |  _ -> [t]
+
+
+(** [get_binary_arrow t] return t' such that t = t' -> t' -> t' *)
+let get_binary_arrow evd t =
+  let l = get_arrow_typ evd t in
+  match l with
+  | []  -> assert false
+  | [t1;t2;t3] ->  Some (t1,t2,t3)
+  |  _   -> None
+
+(** [get_unary_arrow t] return t' such that t = t' -> t' *)
+let get_unary_arrow evd t =
+  let l = get_arrow_typ evd t in
+  match l with
+  | []  -> assert false
+  | [t1;t2] -> Some (t1,t2)
+  |  _   -> None
+
+
+(** [HConstr] is a map indexed by EConstr.t.
+    It should only be used using closed terms.
+ *)
+module HConstr =
+  struct
+    module M =
+      Map.Make(
+          struct
+            type t = EConstr.t
+
+            let compare c c' = Constr.compare (unsafe_to_constr c) (unsafe_to_constr c')
+
+          end)
+
+    let lfind h m =
+      try M.find h m
+      with Not_found -> []
+
+
+    let add h e m =
+      let l = lfind h m in
+      M.add h (e::l) m
+
+    let empty = M.empty
+
+    let find h m =
+      match lfind h m with
+      | e::_ -> e
+      | [] -> raise Not_found
+
+
+    let find_all  = lfind
+
+    let fold f m acc =
+      M.fold (fun k l acc ->
+          List.fold_left (fun acc e -> f k e acc) acc l) m acc
+
+  end
+
+
+(** [get_projections_from_constant (evd,c) ]
+    returns an array of constr [| a1,.. an|] such that [c] is defined as
+    Definition c := mk a1 .. an with mk a constructor.
+    ai is therefore either a type parameter or a projection.
+*)
+let get_projections_from_constant (evd,i) =
+  match EConstr.kind evd i with
+  | Constr.Const(c,_ ) ->
+     begin
+       match Global.body_of_constant Library.indirect_accessor c with
+       | None -> failwith "Add Injection requires a constant (with a body)"
+       | Some(c,_,_) ->
+          match EConstr.kind evd  (EConstr.of_constr c) with
+          | App(c,a) -> Some a
+          |    _            -> None
```

Comment:
Really weird spacing here and other places, maybe run an automatic formatter?

Reply:
Did some best effort at improving formatting.

Reply:
ocamlformat does indeed give great results [once you tweak the flags a bit]

Reply:
done using default flags.

Reply:
Thanks @fajb ! We need to document this better [c.f. #9572] , a set of flags closer to Coq's style is:
```
ocamlformat --module-item-spacing=compact --inplace --break-sequences --sequence-style=terminator --cases-exp-indent=2 --field-space=loose
```